File: backend/app/path_semantics.py
from __future__ import annotations

from pathlib import Path


class PathContractError(ValueError):
    pass


_LEGACY_VAULT_PREFIX = "/vault/"


def normalize_path_separators(raw_path: str | Path) -> str:
    return str(raw_path).replace("\\", "/")


def is_vault_relative(raw_path: str) -> bool:
    try:
        normalized_path = _validate_relative_contract_path(raw_path)
    except PathContractError:
        return False
    return not normalized_path.startswith("/")


def normalize_to_vault_relative(
    raw_path: str | Path,
    *,
    vault_root: Path,
    legacy_mode: bool = False,
) -> str:
    normalized_raw_path = normalize_path_separators(raw_path)
    if not normalized_raw_path.strip():
        raise PathContractError("Path value must be non-empty.")

    resolved_vault_root = vault_root.expanduser().resolve()

    if normalized_raw_path.startswith(_LEGACY_VAULT_PREFIX):
        if not legacy_mode:
            raise PathContractError("Legacy /vault/ paths are not accepted in normal mode.")
        legacy_relative_path = normalized_raw_path.removeprefix(_LEGACY_VAULT_PREFIX)
        if not legacy_relative_path:
            raise PathContractError("Legacy /vault/ path must include a relative target.")
        return _validate_relative_contract_path(legacy_relative_path)

    if normalized_raw_path.startswith("/"):
        _reject_escape_segments(normalized_raw_path)
        resolved_candidate = Path(normalized_raw_path).expanduser().resolve()
        try:
            relative_candidate = resolved_candidate.relative_to(resolved_vault_root)
        except ValueError as exc:
            raise PathContractError("Absolute path must resolve inside the configured vault.") from exc
        return _validate_relative_contract_path(relative_candidate.as_posix())

    return _validate_relative_contract_path(normalized_raw_path)


def resolve_vault_relative(raw_path: str, *, vault_root: Path) -> Path:
    normalized_relative_path = normalize_to_vault_relative(raw_path, vault_root=vault_root)
    return vault_root.expanduser().resolve() / normalized_relative_path


def _validate_relative_contract_path(raw_path: str) -> str:
    normalized_path = normalize_path_separators(raw_path)
    if not normalized_path.strip():
        raise PathContractError("Path value must be non-empty.")
    if normalized_path.startswith("/"):
        raise PathContractError("Path must be vault-relative and not absolute.")
    _reject_escape_segments(normalized_path)
    if any(part == "" for part in normalized_path.split("/")):
        raise PathContractError("Path must not contain empty path segments.")
    if len(normalized_path) >= 2 and normalized_path[1] == ":" and normalized_path[0].isalpha():
        raise PathContractError("Path must be vault-relative and not a drive path.")

    return normalized_path


def _reject_escape_segments(raw_path: str) -> None:
    parts = [part for part in raw_path.split("/") if part]
    if any(part in {".", ".."} for part in parts):
        raise PathContractError("Path must not contain '.' or '..' segments.")

File: plugin/src/writeback/pathSemantics.ts
import path from "path";

const LEGACY_VAULT_PREFIX = "/vault/";

export interface VaultPathApp {
  vault: {
    adapter?: unknown;
  };
}

export class PathContractError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PathContractError";
  }
}

export function normalizeWritebackTargetPath(
  app: VaultPathApp,
  rawPath: string
): string {
  const normalizedRawPath = normalizePathSeparators(rawPath).trim();
  if (normalizedRawPath.length === 0) {
    throw new PathContractError("Path value must be non-empty.");
  }

  if (isWindowsAbsolutePath(normalizedRawPath)) {
    return normalizeWindowsAbsolutePath(app, normalizedRawPath);
  }

  if (normalizedRawPath.startsWith("/")) {
    return normalizePosixAbsolutePath(app, normalizedRawPath);
  }

  return normalizeVaultRelativePath(normalizedRawPath);
}

export function normalizeVaultRelativePath(rawPath: string): string {
  const normalizedRawPath = normalizePathSeparators(rawPath).trim();
  if (normalizedRawPath.length === 0) {
    throw new PathContractError("Path value must be non-empty.");
  }
  if (normalizedRawPath.startsWith("/")) {
    throw new PathContractError("Path must be vault-relative and not absolute.");
  }
  if (isWindowsAbsolutePath(normalizedRawPath)) {
    throw new PathContractError("Path must be vault-relative and not absolute.");
  }
  rejectEscapeSegments(normalizedRawPath);

  const normalizedPath = path.posix.normalize(normalizedRawPath);
  if (
    normalizedPath.length === 0
    || normalizedPath === "."
    || normalizedPath.startsWith("/")
    || normalizedPath === ".."
  ) {
    throw new PathContractError("Path must be vault-relative and not absolute.");
  }
  if (normalizedPath.split("/").some((segment) => segment.length === 0)) {
    throw new PathContractError("Path must not contain empty path segments.");
  }
  return normalizedPath;
}

function normalizePosixAbsolutePath(app: VaultPathApp, rawPath: string): string {
  const adapter = app.vault.adapter as {
    getBasePath?: () => string;
  } | null | undefined;
  if (typeof adapter?.getBasePath !== "function") {
    throw new PathContractError(
      "Absolute target paths require a filesystem-backed Obsidian vault adapter."
    );
  }

  const vaultRoot = normalizeFilesystemPath(path.resolve(adapter.getBasePath()));
  const absolutePath = normalizeFilesystemPath(path.resolve(rawPath));
  if (
    absolutePath !== vaultRoot
    && !absolutePath.startsWith(`${vaultRoot}/`)
  ) {
    if (isLegacyVaultPseudoPath(rawPath)) {
      throw new PathContractError("Legacy /vault/ paths are not accepted in normal mode.");
    }
    throw new PathContractError("Absolute path must resolve inside the configured vault.");
  }

  return normalizeVaultRelativePath(path.relative(vaultRoot, absolutePath));
}

function normalizeWindowsAbsolutePath(app: VaultPathApp, rawPath: string): string {
  const adapter = app.vault.adapter as {
    getBasePath?: () => string;
  } | null | undefined;
  if (typeof adapter?.getBasePath !== "function") {
    throw new PathContractError(
      "Absolute target paths require a filesystem-backed Obsidian vault adapter."
    );
  }

  const vaultRoot = normalizeFilesystemPath(path.win32.resolve(adapter.getBasePath()));
  const absolutePath = normalizeFilesystemPath(path.win32.resolve(rawPath));
  const comparisonVaultRoot = vaultRoot.toLowerCase();
  const comparisonAbsolutePath = absolutePath.toLowerCase();
  if (
    comparisonAbsolutePath !== comparisonVaultRoot
    && !comparisonAbsolutePath.startsWith(`${comparisonVaultRoot}/`)
  ) {
    if (isLegacyVaultPseudoPath(rawPath)) {
      throw new PathContractError("Legacy /vault/ paths are not accepted in normal mode.");
    }
    throw new PathContractError("Absolute path must resolve inside the configured vault.");
  }

  return normalizeVaultRelativePath(path.win32.relative(vaultRoot, absolutePath));
}

function normalizePathSeparators(rawPath: string): string {
  return String(rawPath).replace(/\\/g, "/");
}

function normalizeFilesystemPath(rawPath: string): string {
  return rawPath.replace(/\\/g, "/");
}

function rejectEscapeSegments(rawPath: string): void {
  const parts = rawPath.split("/").filter((part) => part.length > 0);
  if (parts.some((part) => part === "." || part === "..")) {
    throw new PathContractError("Path must not contain '.' or '..' segments.");
  }
  if (rawPath.includes("//")) {
    throw new PathContractError("Path must not contain empty path segments.");
  }
}

function isWindowsAbsolutePath(rawPath: string): boolean {
  return /^[A-Za-z]:\//.test(rawPath) || rawPath.startsWith("//");
}

function isLegacyVaultPseudoPath(rawPath: string): boolean {
  return rawPath === "/vault" || rawPath.startsWith(LEGACY_VAULT_PREFIX);
}
